File: apple/src/ReturnCode.m
/*
 * Copyright (c) 2021, 2026 Taner Sener
 *
 * This file is part of FFmpegKitNext.
 *
 * FFmpegKitNext is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FFmpegKitNext is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General License for more details.
 *
 * You should have received a copy of the GNU Lesser General License
 * along with FFmpegKitNext. If not, see <http://www.gnu.org/licenses/>.
 */

#import "ReturnCode.h"

@implementation ReturnCode {
    int _value;
}

- (instancetype)init:(int)value {
    self = [super init];
    if (self) {
        _value = value;
    }

    return self;
}

+ (BOOL)isSuccess:(ReturnCode *)value {
    return (value != nil) && ([value getValue] == ReturnCodeSuccess);
}

+ (BOOL)isCancel:(ReturnCode *)value {
    return (value != nil) && ([value getValue] == ReturnCodeCancel);
}

- (int)getValue {
    return _value;
}

- (BOOL)isValueSuccess {
    return (_value == ReturnCodeSuccess);
}

- (BOOL)isValueError {
    return ((_value != ReturnCodeSuccess) && (_value != ReturnCodeCancel));
}

- (BOOL)isValueCancel {
    return (_value == ReturnCodeCancel);
}

- (NSString *)description {
    return [NSString stringWithFormat:@"%d", _value];
}

@end

File: windows/src/internal/ReturnCode.cpp
/*
 * Copyright (c) 2022, 2026 Taner Sener
 *
 * This file is part of FFmpegKitNext.
 *
 * FFmpegKitNext is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FFmpegKitNext is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General License for more details.
 *
 * You should have received a copy of the GNU Lesser General License
 * along with FFmpegKitNext. If not, see <http://www.gnu.org/licenses/>.
 */

#include "ReturnCode.h"

bool ffmpegkit::internal::ReturnCode::isSuccess(
    const std::shared_ptr<ffmpegkit::internal::ReturnCode> value) {
    return (value != nullptr) && (value->getValue() == Success);
}

bool ffmpegkit::internal::ReturnCode::isCancel(
    const std::shared_ptr<ffmpegkit::internal::ReturnCode> value) {
    return (value != nullptr) && (value->getValue() == Cancel);
}

ffmpegkit::internal::ReturnCode::ReturnCode(const int value) : _value{value} {}

int ffmpegkit::internal::ReturnCode::getValue() const { return _value; }

bool ffmpegkit::internal::ReturnCode::isValueSuccess() const {
    return (_value == Success);
}

bool ffmpegkit::internal::ReturnCode::isValueError() const {
    return ((_value != Success) && (_value != Cancel));
}

bool ffmpegkit::internal::ReturnCode::isValueCancel() const { return (_value == Cancel); }

namespace ffmpegkit {
namespace internal {

std::ostream &operator<<(std::ostream &out,
                         const std::shared_ptr<ffmpegkit::internal::ReturnCode> &o) {
    if (o == nullptr) {
        return out;
    } else {
        return out << o->_value;
    }
}

} // namespace internal
} // namespace ffmpegkit
